archive/100-theorems-list/90_stirlings_formula.lean

```diff
@@ -0,0 +1,237 @@
+It states that $n!$ grows asymptotically like $\sqrt{2\pi n}(\frac{n}{e})^n$.
+TODO: Add Part 2 to complete the proof
+
+## Proof outline
+
+The proof follows: <https://proofwiki.org/wiki/Stirling%27s_Formula>.
+
+### Part 1
+We consider the fraction sequence $a_n$ of fractions $n!$ over $\sqrt{2n}(\frac{n}{e})^n$ and
+proves that this sequence converges against a real, positve number $a$. For this the two main
+ingredients are
+ - taking the logarithm of the sequence and
+ - use the series expansion of $\log(1 + x)$.
+-/
+
+open_locale topological_space big_operators
+open finset filter nat real
+
+namespace stirling
+
+/-!
+ ### Part 1
+ https://proofwiki.org/wiki/Stirling%27s_Formula#Part_1
+-/
+
+/--
+Define `stirling_seq n` as $\frac{n!}{\sqrt{2n}/(\frac{n}{e})^n$.
+Stirling's formula states that this sequence has limit $\sqrt(π)$.
+-/
+noncomputable def stirling_seq (n : ℕ) : ℝ :=
+n.factorial / (sqrt (2 * n) * (n / exp 1) ^ n)
+
+/-- Define `log_stirling_seq n` as the log of `stirling_seq n`. -/
+noncomputable def log_stirling_seq (n : ℕ) : ℝ := log (stirling_seq n)
+
+/--
+We have the expression
+`log_stirling_seq (n + 1) = log(n + 1)! - 1 / 2 * log(2 * n) - n * log ((n + 1) / e)`.
+-/
+lemma log_stirling_seq_formula (n : ℕ) : log_stirling_seq n.succ =
+  log n.succ.factorial - 1 / 2 * log (2 * n.succ) - n.succ * log (n.succ / exp 1) :=
+begin
+  have h3, from sqrt_ne_zero'.mpr (mul_pos two_pos $ cast_pos.mpr (succ_pos n)),
+  have h4 : 0 ≠ ((n.succ : ℝ) / exp 1) ^ n.succ, from
+    ne_of_lt (pow_pos (div_pos (cast_pos.mpr n.succ_pos ) (exp_pos 1)) n.succ),
+  rw [log_stirling_seq, stirling_seq, log_div, log_mul, sqrt_eq_rpow, log_rpow, log_pow],
+  { linarith },
+  { exact (zero_lt_mul_left zero_lt_two).mpr (cast_lt.mpr n.succ_pos),},
+  { exact h3, },
+  { exact h4.symm, },
+  { exact cast_ne_zero.mpr n.succ.factorial_ne_zero, },
+  { apply (mul_ne_zero h3 h4.symm), },
+end
+
+/--
+The sequence `log_stirling_seq (m + 1) - log_stirling_seq (m + 2)` has the series expansion
+   `∑ 1 / (2 * (k + 1) + 1) * (1 / 2 * (m + 1) + 1)^(2 * (k + 1))`
+-/
+lemma log_stirling_seq_diff_has_sum (m : ℕ) :
+  has_sum (λ k : ℕ, (1 : ℝ) / (2 * k.succ + 1) * ((1 / (2 * m.succ + 1)) ^ 2) ^ k.succ)
+  (log_stirling_seq m.succ - log_stirling_seq m.succ.succ) :=
+begin
+  change
+    has_sum ((λ b : ℕ, 1 / (2 * (b : ℝ) + 1) * ((1 / (2 * m.succ + 1)) ^ 2) ^ b) ∘ succ) _,
+  rw has_sum_nat_add_iff 1,
+  convert (series_log_succ_div $ cast_pos.mpr (succ_pos m)).mul_left ((m.succ : ℝ) + 1 / 2),
+  { ext k,
+    rw [← pow_mul, pow_add],
+    have : 2 * (k : ℝ) + 1 ≠ 0, {norm_cast, exact succ_ne_zero (2*k)},
+    have : 2 * (m.succ : ℝ) + 1 ≠ 0, {norm_cast, exact succ_ne_zero (2*m.succ)},
+    field_simp,
+    ring },
+  { have h : 1 + (m.succ : ℝ)⁻¹ = (m.succ + 1) / (m.succ),
+    { rw add_div _ _ _,
+      rw [one_div, add_left_inj, div_self ],
+      exact nonzero_of_invertible _, },
+    rw h,
```

Comment:
Because of the cahnge in `series_log_succ_div`, I needed to add the hypothesis `h`. I tried a bit to add it into the `simp only` with `discharger` below, but couldn't get that to work.

Reply:
I could do it by making the `h` more general, and then adding it to the `simp only`